### PR TITLE
feat: add integration with Mezzio

### PR DESCRIPTION
**Pull request by DominicDetta**
This pull request allow to use Clockwork  in [Mezzio](https://docs.mezzio.dev/mezzio/) application.

### CODE REVIEW

**Comment by UlrichEckhardt** [comment]
I'm not the maintainer, so just a few non-normative comments:

* Why did you derive from `Clockwork\Support\Vanilla\Clockwork`? First off, the additional utility functions (`isEnabled()` in particular) are really useful additions, but I'd just put them into the baseclass. Then, only the constructor remains, where I find it's really bad practice to not invoke `parent::__construct()`. I think, just using the enhanced vanilla baseclass would work just as well, so that class and config.php can go.
* I think the routing bits in the middleware are not complete yet. There is the `/latest`, a placeholder for the newest record, and there is `/<ID>/extended`, which retrieves full info (in particular the XDebug profiler data) from the according data sources.

That said, cool contribution! I personally don't use Mezzio, but open source lives from participation!

**Comment by DominicDetta** [comment]
Thank you for your fast response.
I derived from `Clockwork\Support\Vanilla\Clockwork` because I want to be able to use all built-in functions but the only down-side was that the original config.php was using the `$_ENV` which I replaced it with a cross-compatible function `getenv`. In this way I guarantee that works in any environment. Furthemore I didnt invoke the `parent::__construct()` because  some application could configure Clockwork directly using environment variables without passing any PHP configuration.
I didnt want to modify the base class but if approved I can move all utility functions to it.
I verified and the `/latest` works correctly. I will try the other missing routing.

**Comment by DominicDetta** [comment]
Do you know if the project is using the  [PSR-12](https://www.php-fig.org/psr/psr-12/) as formatter?
The class `Clockwork\Support\Vanilla\Clockwork` is completely changed

**Comment by DominicDetta** [comment]
I tested also the the routing for `/extended` and it works perfectly.

**Comment by UlrichEckhardt** [comment]
Clockwork seems to use tabs to indent the code. If your IDE autoformatted the files with four-space indent (which is the default per PSR, I believe), then it breaks the formatting here.

**Comment by DominicDetta** [comment]
Yes, I can create a separate package.

The difference between `$_ENV` and `getenv` is obscure to me as well but from my experience in a docker environment with Apache the environment variables are not returned when using `$_ENV`.

**Comment by UlrichEckhardt** [comment]
https://stackoverflow.com/questions/8798294/getenv-vs-env-in-php has some info on that...

**Comment by UlrichEckhardt** [comment]
https://stackoverflow.com/questions/3780866/why-is-my-env-empty may be even better. With that in mind, I'd vote for using `getenv()` exclusively and not using the volatile `$_ENV`. That would be a change for the the Vanilla integration though, not just for Mezzio.

**Comment by DominicDetta** [comment]
To make the integration with clockwork more interesting, may I add that I have the intention of requesting an amendment to [Mezzio](https://docs.mezzio.dev/mezzio/v3/cookbook/debug-toolbars/)'s documentation

File: Clockwork/Support/Mezzio/config.php [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+		// Enable or disable the Web UI, set to the public uri where Clockwork Web UI is accessible
+		'enable' => getenv('CLOCKWORK_WEB_ENABLE') !== false ? getenv('CLOCKWORK_WEB_ENABLE') : true,
+
+		// Path where to install the Web UI assets, should be publicly accessible
+		'path' => getenv('CLOCKWORK_WEB_PATH') !== false ? getenv('CLOCKWORK_WEB_PATH') : 'public/vendor/clockwork',
+
+		// Public URI where the installed Web UI assets will be accessible
+		'uri' => getenv('CLOCKWORK_WEB_URI') !== false ? getenv('CLOCKWORK_WEB_URI') : '/vendor/clockwork',
+
+		// host where web clockwork is accessible, used when a server side request return the html
```
Comment (UlrichEckhardt) [inline Clockwork/Support/Mezzio/config.php:136]: I believe this is what `CLOCKWORK_WEB_URI` is supposed to be.

**Comment by UlrichEckhardt** [comment]
Just wondering, does Mezzio come with a PSR-17 response factory, included or as dependency? Point is, using PSR-7, PSR-15 and PSR-17 as base, I think I could build a middleware that can be integrated into pretty much any framework that uses those HTTP PSRs, including Slim (which is my concern) and Mezzio.

**Comment by DominicDetta** [comment]
When you install Mezzio from the skeleton provided the dependency [laminas-diactoros](https://docs.laminas.dev/laminas-diactoros/) is installed which implement PSR-7 and PSR-17.

File: Clockwork/Support/Mezzio/config.php (hunk 1 above)
Comment (DominicDetta) [reply Clockwork/Support/Mezzio/config.php:136]: As explained by the documentation to serve the WEB UI you need to specify this routing:

**Comment by DominicDetta** [comment]
Right now my middleware use the `Laminas\Diactoros\Response`. If there is a way to create a response without relying to it, it will be completely framework agnostic.

**Comment by itsgoingd** [comment]
@UlrichEckhardt I was actually thinking the other day, why do we still need a custom Slim integration, when the Vanilla integration is superior and more full-featured. I think it would be a really good idea to add a generic PSR-compatible middleware to the Vanilla integration as you suggest.

@DominicDetta I think I would put this on hold and explore the route of adding a generic middleware to support all these PSR-compatible micro-frameworks first.

Also, I'm okay with changing the Vanilla config file to use `getenv`, if anyone wants to make that PR.

File: Clockwork/Support/Mezzio/config.php (hunk 1 above)
Comment (itsgoingd) [reply Clockwork/Support/Mezzio/config.php:136]: The Web UI in the Vanilla integration works in a bit of a funny way. It actually copies the Web UI implementation to a public directory, so the assets are served directly by the web server, this is by default `public/vendor/clockwork`. The code handling the main web UI route, typically `/clockwork`, then returns a special full-screen iframe that embeds the Web UI located at `/vendor/clockwork` with some special configuration to make it work.

Comment (UlrichEckhardt) [reply Clockwork/Support/Mezzio/config.php:136]: This is something that is not really clear and I believe that the different integrations differ a bit, too. Just for reference, here are the two settings I found:
```php
		// Path where to install the Web UI assets, should be publicly accessible
		'path' => isset($_ENV['CLOCKWORK_WEB_PATH']) ? $_ENV['CLOCKWORK_WEB_PATH'] : __DIR__ . '/../../../../../public/vendor/clockwork',

		// Public URI where the installed Web UI assets will be accessible
		'uri' => isset($_ENV['CLOCKWORK_WEB_URI']) ? $_ENV['CLOCKWORK_WEB_URI'] : '/vendor/clockwork'
```
Note that both are commented with the words "accessible" and "public" and both contain a path, which is a bit confusing.

Anyhow, just out of interest, why does the web UI copy the files between paths? I've use the `Web` class' assets and served them from where they are for my PSR-based middleware implementation.

**Comment by UlrichEckhardt** [comment]
https://github.com/UlrichEckhardt/clockwork-slim-demo/tree/feature/psr-17-middleware -- Slim doesn't need a custom integration, it provides implementations for the PSR-{7,15,17} interfaces, so it could use any middleware built on top of that. I've realized that in the above demo project, which should be seen as a proof of concept.

There's still a few things to do:
- `Middleware` needs to be integrated into the clockwork sources, consistently formatted with tabs.
- The configuration access methods need to be integrated into the `Support\Vanilla\Clockwork` class.
- The web path needs to be clarified.
- Documentation.
- Testing. ;)

File: Clockwork/Support/Mezzio/config.php (hunk 1 above)
Comment (DominicDetta) [reply Clockwork/Support/Mezzio/config.php:136]: @itsgoingd 

> The Web UI in the Vanilla integration works in a bit of a funny way. It actually copies the Web UI implementation to a public directory, so the assets are served directly by the web server, this is by default `public/vendor/clockwork`. The code handling the main web UI route, typically `/clockwork`, then returns a special full-screen iframe that embeds the Web UI located at `/vendor/clockwork` with some special configuration to make it work.

This is exactly how I understood when using it. I only added a new env vars `WEB_HOST` to configure the `/clockwork` url but if you think it's useless I can hadcode the url in the middleware and remove it.

**Comment by DominicDetta** [comment]
> @DominicDetta I think I would put this on hold and explore the route of adding a generic middleware to support all these PSR-compatible micro-frameworks first.

Ok, that sounds right.

**Comment by UlrichEckhardt** [comment]
I just added #680, including two demo projects for both Slim and Mezzio. This is still in draft form, in particular there are still a few places marked `TODO: ...` that need work. Still, it's working and something to toy with.

**Comment by DominicDetta** [comment]
This PR will be replaced by #680

**Pull request closed without merging**